lib/Model/Importer/Binary/PECOFFImporter.cpp

```diff
@@ -74,22 +74,48 @@ Error PECOFFImporter::import() {
     }
     const object::coff_section *CoffRef = *SecOrErr;
 
-    // VirtualSize might be larger than SizeOfRawData (extra data at the end of
-    // the section) or viceversa (data mapped in memory but not present in
-    // memory, e.g., .bss)
-    uint64_t SegmentSize = std::min(CoffRef->VirtualSize,
-                                    CoffRef->SizeOfRawData);
-
     MetaAddress Start = ImageBase + u64(CoffRef->VirtualAddress);
     Segment Segment({ Start, u64(CoffRef->VirtualSize) });
 
     Segment.StartOffset = CoffRef->PointerToRawData;
-    Segment.FileSize = Segment.StartOffset + SegmentSize;
+
+    // VirtualSize might be larger than SizeOfRawData (extra data at the end of
+    // the section) or viceversa (data mapped in memory but not present in
+    // memory, e.g., .bss)
+    Segment.FileSize = CoffRef->SizeOfRawData;
+
+    // Since it is possible that the file size if greater than VirtualSize
+    // because SizeOfRawData is rounded, but VirtSize is not,
+    // we work it around here.
+    // TODO: Revisit this.
```

Comment:
Is there something that smells to you about this solution?

Reply:
Hm, it looks like we could be using this as is. I haven't found a lot from the official microsoft documentation about this, but based on some articles from the web [0], if VirtualSize is < then FileSize, the win loader will first round the virt-size up to the memory alignment and then copy the reset of bytes (up to the FileSIze) from File. I am not sure how this all affects the workflow of the upcoming steps from the revng pipeline.

[0] https://stackoverflow.com/questions/44927518/pe-adding-code-at-the-end-of-txt-section

Reply:
Ping